File: java/datastructure/queue/LinkedQueue.java
package datastructure.queue;

/**
 * @author roseduan[roseduan520@gmail.com]
 * @time 2019/10/31 23:30
 * @description 使用链表实现队列
 */
public class LinkedQueue<T> {

    private ListNode<T> head;
    private ListNode<T> tail;
    private int size;

    public LinkedQueue(T val) {
        ListNode<T> node = new ListNode<>(val);
        this.head = node;
        this.tail = node;
        this.size = 1;
    }

    private static class ListNode<T>{
        T val;
        ListNode<T> next;

        ListNode(T val) {
            this.val = val;
            this.next = null;
        }
    }

    /**
     * 入队列
     */
    public boolean add(T value) {
        ListNode<T> node = new ListNode<>(value);
        if (head == null) {
            head = node;
        }
        else {
            tail.next = node;
        }

        tail = node;
        size++;
        return true;
    }

    /**
     * 出队列
     */
    public T dequeue() {
        if (head == null){
            return null;
        }

        ListNode<T> result = head;
        T val = result.val;
        head = head.next;
        result = null;
        size--;

        return val;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        ListNode<T> p = head;
        while (p != null) {
            sb.append(p.val).append(" ");
            p = p.next;
        }

        return sb.toString();
    }
}

File: go/queue/linked_queue/linked_queue.go
package linked_queue

import "fmt"

//链式队列的实现（基于链表）

type ListNode struct {
	next *ListNode
	val  interface{}
}

type LinkedQueue struct {
	head *ListNode
	tail *ListNode
	size int
}

func newNode(val interface{}) *ListNode {
	return &ListNode{nil, val}
}

func New() *LinkedQueue {
	return &LinkedQueue{nil, nil, 0}
}

//入队列
func (q *LinkedQueue) Enqueue(val interface{}) {
	node := newNode(val)

	if q.head == nil {
		q.head = node
	} else {
		q.tail.next = node
	}

	q.tail = node
	q.size++
}

//出队列
func (q *LinkedQueue) Dequeue() *ListNode {
	if q.size <= 0 {
		return nil
	}

	r := q.head
	q.head = q.head.next
	q.size--
	return r
}

//队头节点
func (q *LinkedQueue) Front() *ListNode {
	if q.size <= 0 {
		return nil
	}
	return q.head
}

//队尾节点
func (q *LinkedQueue) Back() *ListNode {
	if q.size <= 0 {
		return nil
	}
	return q.tail
}

func (q *LinkedQueue) Size() int {
	return q.size
}

func (q *LinkedQueue) PrintData() {
	p := q.head
	for p != nil {
		fmt.Print(p.val, " ")
		p = p.next
	}
	fmt.Println()
}
